### PR TITLE
Ast invariants

### PR DESCRIPTION
**Pull request by ghost**
### Overview

Currently the compiler assumes some invariants on the parsetree which are only checked by the typer. As a result it's easy to miss some invariants and we get inconsistent behaviors between the various _clients_ of the parsetree, such as the typer and the output of `-dsource`:
- the typer accepts empty type declarations but the printer rejects them with `assert false`
- the typer rejects empty tuples but the printer prints them as `()`
- the typer accepts `let ... in ...` with no bindings and the printer prints garbage ([PR#7111](http://caml.inria.fr/mantis/view.php?id=7111))
- ...

This patch adds a module `Ast_invariants` collecting all these invariants. In addition to ensuring coherence between various _clients_ of the Parsetree, it formally documents the invariants which are not expressed in the Parsetree definition.
### Implementation details

The check is done by `Pparse`, after applying the ppx rewriters. The second commit replaces the now redundant checks in the typer by asserts.

`Ast_invariants` is internally implemented using an `Ast_mapper.mapper` record. I didn't expose it as I though we could replace it with an `Ast_iter.iter` if we ever add this (it would avoid the extra allocations).

### CODE REVIEW

**Comment by gasche** [comment]
Forbidding empty type declarations seems slightly problematic as the camlp4 revised syntax allows to express them.

Could you measure the (absence of ?) performance impact on bytecode compilation?

**Comment by ghost** [comment]
How do you express empty type declarations with camlp4 revised syntax? I'll look for the performance impact

**Comment by ghost** [comment]
I did this (with ocamlc and ocamlc.opt):

``` sh
for i in 0 10 100 1000 10000 100000; do
  seq -f 'let _ = %g' $i > x.ml
  time ocamlc.opt -c x.ml
done
```

Up to 1000 there was no difference. For 10000 and 100000 it was between 5% and 9% slower.

**Comment by Drup** [comment]
I think this is a good idea, but could we please avoid functions that return unit and raise directly ? I would prefer functions that return something and an auxiliary function that raises.

**Comment by gasche** [comment]
To build an empty type, I use

``` ocaml
type empty = [ ];
```

I just tested and it seems that this is rejected, on the ocaml side, in 4.02.3, but accepted in 4.01 and older OCaml versions. Rejecting this is a bit sad, as I found the abilty to define empty types rather useful for some experiments.

A 5-9% slowdown is about what I would expect for an additional compilation pass. I think this is acceptable as a worst-case scenario (usual compilations will spend more time type-checking and, more importantly, noticeably more time loading compiled interfaces of used libraries).

**Comment by mshinwell** [comment]
I suggested to @diml that he tries using -dtimings on his new pass to get more accurate numbers.

**Comment by ghost** [comment]
> `type empty = [ ];`

Ah, I thought you meant `P(str|sig)_type (_, [])`. For empty variant types the patch adopt the same behavior as trunk. I think that if we want  to accept such types, we should have a syntax for it, otherwise it's not clear what pprintast should do.

**Comment by ghost** [comment]
Using `-dtimings` I got these numbers:

```
ocamlc.opt:
i = 0
all: 0.002s
ast-invariants(blah): 0.000s
typing(x.ml): 0.001s
parsing(x.ml): 0.000s
generate(x.ml): 0.000s
transl(x.ml): 0.000s
i = 10
all: 0.002s
ast-invariants(blah): 0.000s
generate(x.ml): 0.000s
typing(x.ml): 0.000s
transl(x.ml): 0.000s
parsing(x.ml): 0.000s
i = 100
all: 0.003s
ast-invariants(blah): 0.000s
typing(x.ml): 0.001s
parsing(x.ml): 0.000s
generate(x.ml): 0.000s
transl(x.ml): 0.000s
i = 1000
all: 0.009s
parsing(x.ml): 0.004s
ast-invariants(blah): 0.000s
typing(x.ml): 0.003s
generate(x.ml): 0.000s
transl(x.ml): 0.000s
i = 10000
all: 0.089s
parsing(x.ml): 0.038s
ast-invariants(blah): 0.005s
typing(x.ml): 0.037s
transl(x.ml): 0.002s
generate(x.ml): 0.004s
i = 100000
all: 1.185s
parsing(x.ml): 0.440s
ast-invariants(blah): 0.098s
typing(x.ml): 0.535s
transl(x.ml): 0.032s
generate(x.ml): 0.077s

ocamlc:
i = 0
all: 0.004s
ast-invariants(blah): 0.000s
typing(x.ml): 0.000s
transl(x.ml): 0.000s
parsing(x.ml): 0.000s
generate(x.ml): 0.000s
i = 10
all: 0.003s
ast-invariants(blah): 0.000s
generate(x.ml): 0.000s
typing(x.ml): 0.000s
transl(x.ml): 0.000s
parsing(x.ml): 0.000s
i = 100
all: 0.008s
parsing(x.ml): 0.001s
ast-invariants(blah): 0.000s
typing(x.ml): 0.003s
generate(x.ml): 0.000s
transl(x.ml): 0.000s
i = 1000
all: 0.046s
parsing(x.ml): 0.015s
ast-invariants(blah): 0.002s
typing(x.ml): 0.021s
transl(x.ml): 0.001s
generate(x.ml): 0.003s
i = 10000
all: 0.442s
parsing(x.ml): 0.145s
ast-invariants(blah): 0.021s
typing(x.ml): 0.215s
transl(x.ml): 0.012s
generate(x.ml): 0.044s
i = 100000
Fatal error: exception Stack overflow
```

**Comment by alainfrisch** [comment]
> I think that if we want to accept such types, we should have a syntax for it

Same problem for empty pattern matching (or function definitions, I don't remember exactly).  There is no official syntax for it, but this can be expressed in Camlp4 (and as apparently some use cases).

**Comment by gasche** [comment]
(Note that since 4.03 we can express empty matching using refutation clauses: `begin match foo with _ -> . end`.)

**Comment by alainfrisch** [comment]
> (Note that since 4.03 we can express empty matching using refutation clauses: begin match foo with _ -> . end.)

Do you think this could systematically replace the use of (really) empty pattern matching in Camlp4?

**Comment by gasche** [comment]
The semantics is slightly different, as non-exhaustive empty matching is just a warning, while `_ -> .` turns it into an error. So refutation clauses are reasonable when the type really is empty, but would not be suitable if the user used it as a stylistic device that any case is actually non-reachable for non-local reasons. I don't think the latter is a good use-case, though -- it involves silencing exhaustivity warnings which is a very bad idea.

**Comment by ghost** [comment]
Personally I woulnd't bother too much about breaking hidden compiler features that are only accessible via a completely different syntax, but for the sake of not breaking more things, I suggest that we keep the status quo for things such as empty patterns until after 4.03. I.e. the compiler does something and the behavior of -dsource is undefined...

Empty variant types are already not supported by 4.02.3 and trunk so allowing them is independent from this PR.

If there is no objection I'll merge this patch.

**Comment by alainfrisch** [comment]
I would suggest to check invariants after applying each ppx.  This would allow to mention in the error the name of the culprit ppx.  Parsing intermediate binary AST files in the main process might add non-negligible cost, so perhaps the check should be done by each ppx itself (e.g. by calling the check function from Ast_mapper.apply_lazy).  This would also avoid doing the check when there is no ppx at all.

(For "-pp" preprocessors, one could either do the test on their result or decide it is their responsibility to do do.)

**Comment by ghost** [comment]
I think it's better to let the external processes do the invariant checking themselves. Especially since some of them (ppx_deriving, ppx_driver) embeds several transformations. However, to be on the safe side  I think we should always check the invariant  at least once in the compiler if the parsetree doesn't come directly from the parser. To be sure we never get `assert false` in case one uses a non-standard -pp/-ppx.

**Comment by alainfrisch** [comment]
> To be sure we never get assert false in case one uses a non-standard -pp/-ppx.

The extra cost of avoiding those 'assert false' (which would never happen for ppx written using standard drivers, i.e. all of them) is non-negligible, so I'm not sure about the tradeoff.

At least, do you agree that the check should be avoided when no -pp and no -ppx is used?

**Comment by ghost** [comment]
> The extra cost of avoiding those 'assert false' (which would never happen for ppx written using standard drivers, i.e. all of them) is non-negligible, so I'm not sure about the tradeoff.

I just don't feel confortable trusting too much external commands. But I guess we can just say that if people don't follow the standard way it's their fault if bad things happen.

> At least, do you agree that the check should be avoided when no -pp and no -ppx is used?

I was also thinking of the case where the input file is a marshaled AST.

**Comment by gasche** [comment]
Note: I would be interested in getting some version of #351 to go through eventually, and I think the implementation considerations discussed here could maybe be impacted by the new-ish (but minimally interface) it exposes. In the long term, I'm not fond of the current `Pparse` code structure (I find it too complex for what it does), and I would be interested in some deeper refactoring. In any case feedback on #351 would be warmly welcome.

**Comment by damiendoligez** [comment]
> I was also thinking of the case where the input file is a marshaled AST.

This is morally a sub-case of "`-pp` was used", so I agree with you.

File: parsing/ast_invariants.ml [hunk 1]
```diff
@@ -0,0 +1,93 @@
+(***********************************************************************)
+(*                                                                     *)
+(*                                OCaml                                *)
+(*                                                                     *)
+(*                 Jeremie Dimino, Jane Street Europe                  *)
```
Comment (damiendoligez) [inline parsing/ast_invariants.ml:5]: AFAIK, this should say "Jane Street Group", check with @mshinwell .

**Comment by damiendoligez** [comment]
@diml could you change the code to check only when the AST doesn't come from the parser?

@alainfrisch what about a command-line flag to suppress the check?

**Comment by alainfrisch** [comment]
> @alainfrisch what about a command-line flag to suppress the check?

I don't see the point.  People using -pp or -ppx probably don't care about the tiny runtime overhead anyway.  For other people, one should just disable the check automatically.

**Comment by ghost** [comment]
@damiendoligez, sure I pushed the commits. I also added some tests, fixed the headers and added  invariants about where Lapply is allowed

@alainfrisch, I looked at adding the invariant checking in `Ast_mapper` but then there is an annoying circular dependency between `Ast_mapper` and `Ast_invariants`. I was thinking to rearrange things a bit or add `Ast_iter` which would also improve perfs, but I don't think I'll have time to look at this soon.

I also looked at adding checks for constants as suggested in #402 (by just adding `Lexer.is_{int,float}_literal : Lexing.lexbuf -> bool` to reuse the same regexps) but I haven't looked at timings so I left that out for now.

I'm in favor of merging this now as it replaces a few `assert false` and `Misc.fatal_error` by proper errors and leave further improvements for post 4.03

File: testsuite/tests/ast-invariants/test.ml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+(*  the GNU Lesser General Public License version 2.1, with the        *)
+(*  special exception on linking described in the file ../LICENSE.     *)
+(*                                                                     *)
+(***********************************************************************)
+
+(* This test checks all ml files in the ocaml repository that are accepted
+   by the parser satisfy [Ast_invariants].
+
+   We don't check the invariants on the output of the parser, so this test
+   is to ensure that we the parser doesn't accept more than [Ast_invariants].
```
Comment (alainfrisch) [inline testsuite/tests/ast-invariants/test.ml:19]: Typo ("we").

**Comment by alainfrisch** [comment]
The new test (in testsuite/tests/ast-invariants) is broken:

```
Running tests from 'tests/ast-invariants' ...
File "test.ml", line 26, characters 13-32:
Error: Unbound module Parsetree
/home/frisch/ocaml/trunk/testsuite/makefiles/Makefile.common:115: recipe for target 'test.cmx' failed
make[4]: *** [test.cmx] Error 2
File "_none_", line 1:
```

(Note: the testsuite is executed from a source tree before the "make install".)

**Comment by alainfrisch** [comment]
The problem was specific to Windows (Cygwin paths passed to the compiler).  I have pushed a tentative fix.

**Pull request merged**